subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -352,183 +274,159 @@ static void device_twin_request(void)
-		break;
+	if (az_span_size(method_request->request_id) == 0) {
+		LOG_WRN("No request ID, direct method processing aborted");
+		return false;
 	}
 
-	azure_iot_hub_notify_event(&evt);
-}
+	evt.data.method.name.ptr = az_span_ptr(method_request->name);
+	evt.data.method.name.size = az_span_size(method_request->name);
 
-/* Returns true if event should not be propagated from the calling function */
-static bool on_suback(void)
-{
-#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS)
-	if (dps_reg_in_progress()) {
-		int err = dps_send_reg_request();
+	evt.data.method.request_id.ptr = az_span_ptr(method_request->request_id);
+	evt.data.method.request_id.size = az_span_size(method_request->request_id);
 
-		if (err) {
-			LOG_ERR("DPS registration not sent, error: %d",
-				err);
-		} else {
-			LOG_DBG("DPS registration request sent");
-		}
+	LOG_DBG("Direct method name: %.*s",
+		evt.data.method.name.size,
+		log_strdup(evt.data.method.name.ptr));
 
-		return true;
-	}
-#endif /* IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS) */
+	LOG_DBG("Direct method request ID: %.*s",
+		evt.data.method.request_id.size,
+		log_strdup(evt.data.method.request_id.ptr));
 
-	if (IS_ENABLED(CONFIG_AZURE_IOT_HUB_AUTO_DEVICE_TWIN_REQUEST) ||
-	    IS_ENABLED(CONFIG_AZURE_FOTA)) {
-		device_twin_request();
-	}
+	azure_iot_hub_notify_event(&evt);
 
-	return false;
+	return true;
 }
 
-static void on_publish(struct mqtt_client *const client,
-			   const struct mqtt_evt *mqtt_evt)
+AZ_HUB_STATIC void on_publish(struct azure_iot_hub_buf topic, struct azure_iot_hub_buf payload)
 {
-	int err;
-	const struct mqtt_publish_param *p = &mqtt_evt->param.publish;
-	size_t payload_len = p->message.payload.len;
-	struct azure_iot_hub_prop_bag prop_bag[TOPIC_PROP_BAG_COUNT];
+	struct azure_iot_hub_property properties[CONFIG_AZURE_IOT_HUB_MSG_PROPERTY_RECV_MAX_COUNT];
 	struct azure_iot_hub_evt evt = {
 		.type = AZURE_IOT_HUB_EVT_DATA_RECEIVED,
-		.data.msg.ptr = payload_buf,
-		.data.msg.len = payload_len,
-		.topic.str = (char *)p->message.topic.topic.utf8,
-		.topic.len = p->message.topic.topic.size,
-		.topic.prop_bag = prop_bag,
-	};
-	struct topic_parser_data topic_data = {
-		.topic = p->message.topic.topic.utf8,
-		.topic_len = p->message.topic.topic.size,
-		.type = TOPIC_TYPE_UNKNOWN,
+		.data.msg.payload = payload,
+		.topic.ptr = topic.ptr,
+		.topic.size = topic.size,
+		.topic.properties = properties,
 	};
+	az_span topic_span = az_span_create(topic.ptr, topic.size);
+	az_iot_hub_client_c2d_request c2d_msg;
+	az_iot_hub_client_method_request method_request;
+	az_iot_hub_client_twin_response twin_msg;
 
-	LOG_DBG("MQTT_EVT_PUBLISH: id = %d, len = %d ",
-		p->message_id, payload_len);
-
-	err = publish_get_payload(client, payload_len);
-	if (err) {
-		LOG_ERR("publish_get_payload, error: %d", err);
+	/* The topic needs to be parsed with multiple functions in order
+	 * to determine the type of incoming message.
+	 */
+	if (az_result_succeeded(
```

Comment:
This function (`on_publish`) is over 100 lines long. Looks like this whole if-check is used to get the topic type. If it was extracted to it's own function that would cut ~50 lines from `on_publish`.